File: solutions/ruby/say/1/say.rb
class Say
  THOUSAND_LABEL = [nil, 'thousand', 'million', 'billion']
  NUMBERS = {
    0 => 'zero',
    1 => 'one',
    2 => 'two',
    3 => 'three',
    4 => 'four',
    5 => 'five',
    6 => 'six',
    7 => 'seven',
    8 => 'eight',
    9 => 'nine',
    10 => 'ten',
    11 => 'eleven',
    12 => 'twelve',
    13 => 'thirteen',
    20 => 'twenty',
    30 => 'thirty',
    40 => 'forty',
    50 => 'fifty',
    60 => 'sixty',
    70 => 'seventy',
    80 => 'eighty',
    90 => 'ninety',
  }.freeze


  def initialize(number)
    raise ArgumentError.new() if number < 0 || number >= 1_000_000_000_000

    @number = number
  end

  def in_english
    number = @number.to_s
    len = number.length
    number = number.rjust(len + (3 - len) % 3, '0')

    number.scan(/.../).reverse.map.with_index{|block, index|
      next if block.to_i.zero? and (!index.zero? || number.length > 3)
      [tr_hundred(block.to_i), THOUSAND_LABEL[index]].compact.join(' ')
    }.reverse.join(' ').strip
  end

  private
  def tr_hundred(number)
    hundred = number / 100
    number %= 100
    ten = number / 10
    unit = number % 10

    hundred = nil if hundred.zero?
    ten = nil if ten.zero?
    unit = nil if unit.zero? && (!ten.nil? || !hundred.nil?)

    is_special_ten = !unit.nil? && !ten.nil? && !NUMBERS[ten * 10 + unit].nil?

    hundred_str = "#{NUMBERS[hundred]} hundred" unless hundred.nil?
    ten_str = NUMBERS[ten * 10 + unit] if is_special_ten # 11 / 12 / 13 cases
    ten_str ||= NUMBERS[ten * 10] unless ten.nil?
    unit_str = NUMBERS[unit] unless unit.nil? || is_special_ten

    if ten_str == NUMBERS[10] && !unit_str.nil? # teen cases
      unit_str = "#{unit_str}teen"
      ten_str = nil
    end

    [hundred_str, [ten_str, unit_str].compact.join('-')].compact.join(' ')
  end
end

File: solutions/python/say/1/say.py
import functools


def generate_teen(number: int) -> str:
    return f'{numbers[number]}teen'


def generate_thenth(prefix: str, number: int) -> str:
    return f'{prefix}-{numbers[number]}'


numbers = {
    0: 'zero',
    1: 'one',
    2: 'two',
    3: 'three',
    4: 'four',
    5: 'five',
    6: 'six',
    7: 'seven',
    8: 'eight',
    9: 'nine',
    10: 'ten',
    11: 'eleven',
    12: 'twelve',
    13: 'thirteen',
    20: 'twenty',
    30: 'thirty',
    40: 'forty',
    50: 'fifty',
    60: 'sixty',
    70: 'seventy',
    80: 'eighty',
    90: 'ninety',
}


def generate_thenth_dict(number: int) -> dict:
    return dict(zip(range(number+1, number+10), map(lambda i: generate_thenth(numbers[number], i), range(1, 10))))


def merge_dicts(dicts: list[dict]) -> dict:
    return functools.reduce(lambda acc, value: {**acc, **value}, dicts, {})


numbers = {
    **numbers,
    **dict(zip(range(14, 20), map(generate_teen, range(4, 10)))),
    **merge_dicts(map(generate_thenth_dict, range(20, 100, 10)))
}


def chuncks(string: str, split_n: int) -> list[str]:
    return [string[i:i+split_n] for i in range(0, len(string), split_n)]


def rev(string: str):
    return string[::-1]


def say_group(group_number: str) -> str:
    hundredth, tenth, unit = map(int, group_number.rjust(3, '0'))
    if all(map(lambda x: x == 0, [hundredth, tenth, unit])):
        return numbers[0]

    hundredth = f"{numbers[hundredth]} hundred" if hundredth else None
    tenth = tenth * 10 + unit
    tenth = numbers[tenth] if tenth else None

    return ' '.join([str(group) for group in [hundredth, tenth] if group is not None])


SIZE = [None, 'thousand', 'million', 'billion']


def say(number: int) -> str:
    if not 0 <= number < 1000000000000:
        raise ValueError('input out of range')

    *groups, = map(rev, chuncks(str(number)[::-1], 3)[::-1])

    groups_str = []
    for index, group in enumerate(groups):
        suffix = SIZE[len(groups) - index - 1]
        group_str = say_group(group)
        # Case when having a 'zero' in the middle of the groups
        if group_str == numbers[0] and suffix is not None:
            continue
        # Case when having a 'zero' at the end, but having some results before, the 'zero' is absent
        if suffix is None and group_str == numbers[0] and len(groups_str) > 0:
            continue
        groups_str.append(group_str)
        if suffix:
            groups_str.append(suffix)

    return ' '.join(groups_str)
